Improve readability and completedness for Redis

martin-rueegg opened the pull request:
- give more complete examples for Redis config
- use separate sections for the different connection types (TCP, TLS, socket)

Signed-off-by: Martin Rüegg <martin.rueegg@metaworx.ch>

blizzz commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -87,32 +87,120 @@ You can verify that the Redis daemon is running with ``ps ax``::
 
+Further more, you can use Redis for the distributed server cache:: 
+
+  'memcache.distributed' => '\OC\Memcache\Redis',
```
It's not furthermore, but the essential configuration

martin-rueegg replied:
Why essential? You could use _memcached_ for the distributed cache instead (`'memcache.distributed' => '\OC\Memcache\Memcached'`), and _Redis_ only for local cache. As such, _Redis_ is just one available option for the distributed cache. That's why I used "further more".

martin-rueegg replied:
After reading your next comment, it appears that using Redis for local cache only doesn't make much sense. I wonder if that would not need to be explained in some part of the document. I.e. recommended setups are:

- APCu for local cache
- Redis for distributed cache and file locking on a single server
- Redis cluster or Memcached cluster for muliti-server setups

Or something along those lines.

blizzz replied:
> Why essential? You could use _memcached_ for the distributed cache instead (`'memcache.distributed' => '\OC\Memcache\Memcached'`), and _Redis_ only for local cache. As such, _Redis_ is just one available option for the distributed cache. That's why I used "further more".

For me it sounded that the `'memcache.distributed'` option is a nice to have.

blizzz commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -87,32 +87,120 @@ You can verify that the Redis daemon is running with ``ps ax``::
+
+  'memcache.distributed' => '\OC\Memcache\Redis',
+
+Additionally, you can use Redis for the local cache like so:: 
+
+  'memcache.local' => '\OC\Memcache\Redis',
```
You can do it on single-server systems only, with more than app servers it can cause trouble. Even on single-server setups APCu should be faster. So this is possible, but not recommendable.

martin-rueegg replied:
I've incorporated your comments in a note.

blizzz reviewed (dismissed):
Great improvement, few nitpicks :)

martin-rueegg commented:
Thanks @blizzz for taking the time for reviewing this.

nickvergessen commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -87,32 +87,122 @@ You can verify that the Redis daemon is running with ``ps ax``::
+   ]
+
+
+Connecting to Single Redis Server over TCP
```
```suggestion
Connecting to single Redis server over TCP
```

CarlSchwan commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -87,32 +87,122 @@ You can verify that the Redis daemon is running with ``ps ax``::
 
   'memcache.locking' => '\OC\Memcache\Redis',
 
+Further more, you can use Redis for the distributed server cache:: 
+
+  'memcache.distributed' => '\OC\Memcache\Redis',
```
```suggestion
Additionally, you should use Redis for the distributed server cache:: 

  'memcache.distributed' => '\OC\Memcache\Redis',
```

martin-rueegg commented:
@nickvergessen, @blizzz, @CarlSchwan are you happy with the changes I made following your request?

nickvergessen commented:
I'm fine but I have little to no background to verify the content of your changes

martin-rueegg commented:
@PVince81 @skjnldsv , any chance to have a look at this, please?

szaimen commented:
@blizzz can you approve? :)

blizzz commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -30,19 +29,79 @@ of memcache that best fits your needs. The supported caching backends are:
+
+Only use APCu::
+
+    'memcache.local' => '\OC\Memcache\APCu',
```
also set distributed, because both may be used, and the use case is a different one.

martin-rueegg replied:
@blizzz also this one has been there [before](https://github.com/nextcloud/documentation/pull/9093/files#diff-48dc187eeb03371b5ad6886b33043bf060884809c1d51a3568c5d29e40e3bf13L221-L227).

blizzz commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -30,19 +29,79 @@ of memcache that best fits your needs. The supported caching backends are:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+APCu is faster at local caching than Redis. If you have enough memory, use APCu for Memory Caching
+and Redis for File Locking. If you are low on memory, use Redis for both.
```
That probably opens up questions on "why" and also wat "low memory" actually means.

martin-rueegg replied:
This is not new, I've just moved it to a different place: see [current doc](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#additional-notes-for-redis-vs-apcu-on-memory-caching)

martin-rueegg replied:
Also here. I think it would be great, if we could go ahead with this PR, or assign someone who is capable to technically review the content. Something which _I_ am not capable of. What do you say, @blizzz?

blizzz commented on admin_manual/configuration_server/caching_configuration.rst:
```diff
@@ -149,23 +300,31 @@ prevent session corruption when using Redis as your session handler: ::
+Additional Redis installation help
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If your version of **Mint** or **Ubuntu** does not package the required version of 
+``php-redis``, then try `this Redis guide on Tech and Me
+<https://www.techandme.se/install-redis-cache-on-ubuntu-server-with-php-7-and-nextcloud/>`_
+for a complete Redis installation on Ubuntu 14.04 using PECL. 
+These instructions are adaptable for **any distro** that does not package the 
+supported version, or that does not package Redis at all, such as **SUSE Linux 
+Enterprise Server** and **Red Hat Enterprise Linux**.
+
+For PHP 7.0 and PHP 7.1 use Redis PHP module 3.1.x or later.
```
I think we can spare us this section, for 7.4 is the minimum requirement at the moment.

martin-rueegg replied:
This is not my stuff. OF course it can be removed, but it has been there [before](ation/pull/9093/files#diff-48dc187eeb03371b5ad6886b33043bf060884809c1d51a3568c5d29e40e3bf13L249).

If you want me to remove it, please specify exactly which part (lines). But it is a shame, if the improvements of this PR are solely delayed due to old stuff. And honestly, I do not know enough about these technical details to be able to "write" the manual. I've only re-organized and added more examples and explanations ... :-)

martin-rueegg commented:
@PVince81 @szaimen @skjnldsv 

All points raised by @blizzz regard information/sections, that have already been there before. I am not able to technically review the content of the entire document. But if additional changes are required then I suggest to proceed as follows:
- If the change regards the formatting or layout or structure of the article, I'm happy to adjust those.
- If the change regards the technical content, then a new PR/issue should be opened to address those.

Otherwise, we keep on the improvements in this PR from being published, and potentially helping users/admins. Just because the document should have been overhauled/redacted technically already.

What do you guys say?

martin-rueegg commented:
Thanks, @skjnldsv.

skjnldsv commented:
Thanks for your patience @martin-rueegg :hugs:

ChristophWurst commented:
I found 

> Using Redis for local cache on a multi-server setup can cause issues. Also, even on a single-server setup, APCu (see section above) should be faster.

for https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#redis-configuration-in-nextcloud-config-php but not https://docs.nextcloud.com/server/25/admin_manual/configuration_server/caching_configuration.html. Does that mean with Nextcloud >= using Redis for local cache is not problematic? Otherwise I would suggest to backport the refinements of our recommendation.

ChristophWurst commented:
My remark is specifically about b502cc7f27b8cc332401500b42655cc0f377faa9. We can also only backport this one commit and leave the restructuring for 26+.

martin-rueegg commented:
> My remark is specifically about [b502cc7](https://github.com/nextcloud/documentation/commit/b502cc7f27b8cc332401500b42655cc0f377faa9). We can also only backport this one commit and leave the restructuring for 26+.

It's funny, because when I created this PR back in July last year, versions 22-24 where actively released. Don't think any of the examples I've added have been unavailable at that point. So why would it not be back ported to at least v24 or even before?

The note you mentioned, I've added it due to @blizzz 's [comment](https://github.com/nextcloud/documentation/pull/9093#discussion_r928007746).
